src/HotChocolate/ApolloFederation/src/ApolloFederation/FederationSchemaPrinter.cs

```diff
@@ -0,0 +1,398 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using HotChocolate.Language;
+using HotChocolate.Types;
+using HotChocolate.Types.Introspection;
+
+namespace HotChocolate.ApolloFederation
+{
+    public static class FederationSchemaPrinter
+    {
+        public static string Print(ISchema schema)
+        {
+            if (schema is null)
+            {
+                throw new ArgumentNullException(nameof(schema));
+            }
+
+            return SerializeSchema(schema).ToString();
+        }
+
+        private static DocumentNode SerializeSchema(
+            ISchema schema)
+        {
+            var referenced = new ReferencedTypes();
+
+            var typeDefinitions = GetNonScalarTypes(schema)
+                .Select(
+                    t => SerializeNonScalarTypeDefinition(
+                        t,
+                        referenced))
+                .Where(node => node != null)
+                .OfType<IDefinitionNode>()
+                .ToList();
+
+            return new DocumentNode(
+                null,
+                typeDefinitions);
+        }
+
+        private static IEnumerable<INamedType> GetNonScalarTypes(
+            ISchema schema)
+        {
+            return schema.Types
+                .Where(t => IsPublicAndNoScalar(t))
+                .Where(t => !IsApolloTypeAddition(t))
+                .OrderBy(
+                    t => t.Name.ToString(),
+                    StringComparer.Ordinal
+                )
+                .GroupBy(t => (int)t.Kind)
+                .OrderBy(t => t.Key)
+                .SelectMany(t => t);
+        }
+
+        private static bool IsApolloTypeAddition(INamedType type)
+        {
+            if (type is EntityType || type is ServiceType)
```

Comment:
I think we can just return the boolean here :)

Reply:
Oops :D Fixed